src/lib/coreaudio/CoreAudioDevice.cxx

```diff
@@ -0,0 +1,676 @@
+		return channels;
+  
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyStreamConfiguration;
+
+	UInt32 size = 0;
+	OSStatus err = AudioObjectGetPropertyDataSize(device_id, &aopa, 0, NULL, &size);
+	if (err != noErr)
+		return channels;
+  
+	AudioBufferList* buffer_list = (AudioBufferList*)malloc(size);
+	err = AudioObjectGetPropertyData(device_id, &aopa, 0, NULL, &size, buffer_list);
+	if (err == noErr)
+	{
+		if (stream_idx < buffer_list->mNumberBuffers)
+			channels = buffer_list->mBuffers[stream_idx].mNumberChannels;
+	}
+	else
+	{
+	  FormatError(macos_output_domain, "Unable to get number of stream output channels - id: 0x%04x. Error = %s.",
+				  (uint)device_id, GetError(err).c_str());
+	}
+  
+	free(buffer_list);
+  
+	return channels;
+}
+
+bool
+CoreAudioDevice::GetStreams(AudioStreamIdList* stream_id_list)
+{
+	if (!stream_id_list || !device_id)
+		return false;
+
+	// Get only output streams (mScope)
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyStreams;
+
+	UInt32  size = 0;
+	OSStatus err = AudioObjectGetPropertyDataSize(device_id, &aopa, 0, NULL, &size);
+	if (err != noErr)
+		return false;
+
+	UInt32 stream_count = size / sizeof(AudioStreamID);
+	AudioStreamID* streams_list = new AudioStreamID[stream_count];
+	
+	err = AudioObjectGetPropertyData(device_id, &aopa, 0, NULL, &size, streams_list);
+	if (err == noErr)
+	{
+		for (UInt32 stream = 0; stream < stream_count; stream++)
+			stream_id_list->push_back(streams_list[stream]);
+	}
+	
+	delete[] streams_list;
+	return err == noErr;
+}
+
+bool
+CoreAudioDevice::SetHogStatus(bool hog)
+{
+  /* According to Jeff Moore (Core Audio, Apple), Setting kAudioDevicePropertyHogMode
+   is a toggle and the only way to tell if you do get hog mode is to compare
+   the returned pid against getpid, if the match, you have hog mode, if not you don't. */
+	if (!device_id)
+		return false;
+
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyHogMode;
+
+	if (hog)
+	{
+		// Not already set
+		if (hog_pid == -1)
+		{
+			OSStatus err = AudioObjectSetPropertyData(device_id, &aopa, 0, NULL, sizeof(hog_pid), &hog_pid);
+
+			/* Even if setting hogmode was successful our PID might not get written
+			 into hog_pid (so it stays -1). Readback hogstatus for judging if we
+			 had success on getting hog status
+			 We do this only when AudioObjectSetPropertyData didn't set hog_pid because
+			 it seems that in the other cases the GetHogStatus could return -1
+			 which would overwrite our valid hog_pid again */
+			if (hog_pid == -1)
+				hog_pid = GetHogStatus();
+
+			if (err || hog_pid != getpid())
+			{
+				FormatError(macos_output_domain, "Unable to set hog mode. Error = %s", GetError(err).c_str());
```

Comment:
Throw exception?

Reply:
OK, if hog mode cannot be set it means that some other program already owns hog mode which is in any case fatal for using the device further...